File: piece.py
import re

SPACE = " "

def castle(move, board_view, piece_view):
    if move in "OOO":
        home_rank, king, rook = "1", "K", "R"
    else:
        home_rank, king, rook = "8", "k", "r"
    king_before = "e" + home_rank
    if len(move) == 2:
        rook_before = "h" + home_rank
        king_after = "g" + home_rank
        rook_after = "f" + home_rank
    else:
        rook_before = "a" + home_rank
        king_after = "c" + home_rank
        rook_after = "d" + home_rank
    
    board_view[king_before] = SPACE
    board_view[rook_before] = SPACE
    board_view[king_after] = king
    board_view[rook_after] = rook
    piece_view[king].append(king_after)
    piece_view[king].remove(king_before)
    piece_view[rook].append(rook_after)
    piece_view[rook].remove(rook_before)
    
    return board_view, piece_view

def not_blocked(from_sq, to_sq, board_view):
    if from_sq[0] == to_sq[0]:
        between = [from_sq[0] + x for x in "12345678" if (int(from_sq[1]) < int(x) < int(to_sq[1]) or int(to_sq[1]) < int(x) < int(from_sq[1]))]
    elif from_sq[1] == to_sq[1]:
        between = [x + from_sq[1] for x in "abcdefgh" if (ord(from_sq[0]) < ord(x) < ord(to_sq[0]) or ord(to_sq[0]) < ord(x) < ord(from_sq[0]))]
    elif ord(from_sq[0]) < ord(to_sq[1]) and int(from_sq[1]) < int(to_sq[1]):
        between = [chr(x) + str(y) for x, y in zip(range(ord(from_sq[0]) + 1, ord(to_sq[0])), range(int(from_sq[1]) + 1, int(to_sq[1])))]
    elif ord(from_sq[0]) > ord(to_sq[1]) and int(from_sq[1]) < int(to_sq[1]):
        between = [chr(x) + str(y) for x, y in zip(range(ord(to_sq[0]) + 1, ord(from_sq[0])), range(int(to_sq[1]) - 1, int(from_sq[1]) - 2, -1))]
    elif ord(from_sq[0]) < ord(to_sq[1]) and int(from_sq[1]) > int(to_sq[1]):
        between = [chr(x) + str(y) for x, y in zip(range(ord(to_sq[0]) - 1, ord(from_sq[0]) -2, -1), range(int(to_sq[1]) + 1, int(from_sq[1])))]
    elif ord(from_sq[0]) > ord(to_sq[1]) and int(from_sq[1]) > int(to_sq[1]):
        between = [chr(x) + str(y) for x, y in zip(range(ord(to_sq[0]) + 1, ord(from_sq[0])), range(int(to_sq[1]) + 1, int(from_sq[1])))]
    for sq in between:
        if board_view[sq] != " ":
            return False
    return True

def get_from_sq(move, board_view, piece_view):
    piece = move[0]
    all_locations = piece_view[piece]
    possible_locations = []
    for sq in all_locations:
        if check_move(piece, sq, move[-2:]):
            possible_locations.append(sq)
    if len(possible_locations) == 1:
        return possible_locations[0]
    limit_possible_locations = []
    if piece not in "Nn":
        for sq in possible_locations:
            if not_blocked(sq, move[-2:], board_view):
                limit_possible_locations.append(sq)
        if len(limit_possible_locations) == 1:
            return limit_possible_locations[0]
        else:
            return remove_ambiguity(move, limit_possible_locations)
    else:
        return remove_ambiguity(move, possible_locations)

def remove_ambiguity(move, possibilities): 
    if re.search("[a-h][1-8]", move[1:-2]):
        location = re.findall("[a-h][1-8]", move[1:-2])[0]
    elif re.search("[a-h]", move[1:-2]):
        for i in possibilities:
            if re.findall("[a-h]", move[1:-2])[0] in i:
                location = i
                break
    elif re.search("[1-8]", move[1:-2]):
        for i in possibilities:
            if re.findall("[1-8]", move[1:-2])[0] in i:
                location = i
                break
    return location

def check_move(piece, location, destination):
    moved_by = abs(ord(location[0]) - ord(destination[0])), abs(ord(location[1]) - ord(destination[1]))
    
    def can_rook_move():
        return moved_by[0] == 0 or moved_by[1] == 0

    def can_knight_move():
        return moved_by in [(2, 1), (1, 2)]

    def can_bishop_move():
        return moved_by[0] == moved_by[1]

    def can_queen_move():
        return can_rook_move() or can_bishop_move()

    def can_king_move():
        return moved_by in [(0, 1), (1, 0), (1, 1)]

    return {"R": can_rook_move, "N": can_knight_move, "B": can_bishop_move, "Q": can_queen_move, "K": can_king_move}[piece.upper()]()

def move_piece(move, board_view, piece_view):
    from_sq = get_from_sq(move, board_view, piece_view)
    to_sq = move[-2:]
    
    if "x" in move:
        piece_view[board_view[to_sq]].remove(to_sq)

    piece_view[move[0]].append(to_sq)
    piece_view[move[0]].remove(from_sq)
    board_view[from_sq] = SPACE
    board_view[to_sq] = move[0]

    return board_view, piece_view

File: src/Piece.kt
import kotlin.math.absoluteValue

const val SPACE: String = " "

fun castle(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    val homeRank: String
    val king: String
    val rook: String
    if (move in "OOO") {
        homeRank = "1"
        king = "K"
        rook = "R"
    }
    else{
        homeRank = "8"
        king = "k"
        rook = "r"
    }

    val kingBefore = "e$homeRank"
    val rookBefore: String
    val kingAfter: String
    val rookAfter: String
    if (move.length == 2) {
        rookBefore = "h$homeRank"
        kingAfter = "g$homeRank"
        rookAfter = "f$homeRank"
    }
    else {
        rookBefore = "a$homeRank"
        kingAfter = "c$homeRank"
        rookAfter = "d$homeRank"
    }
    boardView[kingBefore] = SPACE
    boardView[rookBefore] = SPACE
    boardView[kingAfter] = king
    boardView[rookAfter] = rook
    pieceView[king]!!.add(kingAfter)
    pieceView[king]!!.remove(kingBefore)
    pieceView[rook]!!.add(rookAfter)
    pieceView[rook]!!.remove(rookBefore)
    return Board(boardView, pieceView)
}

fun notBlocked(fromSq: String, toSq: String, boardView: HashMap<String, String>): Boolean {
    val fromFile: Char = fromSq[0]
    val toFile: Char = toSq[0]
    val fromRank: Int = fromSq[1].toString().toInt()
    val toRank: Int = toSq[1].toString().toInt()
    var noBlocking: Boolean = true
    if (fromFile == toFile && fromRank < toRank){
        for (rank in fromRank+1 until toRank){
            if (boardView[fromFile + rank.toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile == toFile && fromRank > toRank){
        for (rank in toRank+1 until fromRank){
            if (boardView[fromFile + rank.toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile > toFile && fromRank == toRank){
        for (file in toFile+1 until fromFile){
            if (boardView[file + fromRank.toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile < toFile && fromRank == toRank){
        for (file in fromFile+1 until toFile){
            if (boardView[file + fromRank.toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile < toFile && fromRank < toRank){
        for (diff in 1..toRank-fromRank){
            if (boardView[(fromFile.toInt()+diff).toChar() + (fromRank+diff).toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile < toFile && fromRank > toRank){
        for (diff in 1..fromRank-toRank){
            if (boardView[(fromFile.toInt()+diff).toChar() + (fromRank-diff).toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile > toFile && fromRank < toRank){
        for (diff in 1 until toRank-fromRank){
            if (boardView[(fromFile.toInt()-diff).toChar() + (fromRank+diff).toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    else if (fromFile > toFile && fromRank > toRank){
        for (diff in 1..fromRank-toRank){
            if (boardView[(fromFile.toInt()-diff).toChar() + (fromRank-diff).toString()] != SPACE){
                noBlocking = false
                break
            }
        }
    }
    return noBlocking
}

fun removeAmbiguity(move: String, possibilities: MutableList<String>): String{
    val extra: String = move.slice(1..move.length-3)
    var location: String = SPACE
    if (Regex("[a-h][1-8]").containsMatchIn(extra)){
        location = (Regex("[a-h][1-8]").find(extra)?.value).toString()
    }
    else if(Regex("[a-h]").containsMatchIn(extra)){
        for (i in possibilities){
            if (((Regex("[a-h]").find(extra))?.value).toString() in i){
                location = i
            }
        }
    }
    else if(Regex("[1-8]").containsMatchIn(extra)){
        for (i in possibilities){
            if (((Regex("[1-8]").find(extra))?.value).toString() in i){
                location = i
            }
        }
    }
    return location
}

fun checkMove(piece: String, location: String, destination: String): Boolean{
    val movedBy: List<Int> = listOf((location[0]-destination[0]).toInt().absoluteValue, (location[1]-destination[1]).toInt().absoluteValue)

    fun canRookMove(): Boolean{
        return ((movedBy[0] == 0) || (movedBy[1] == 0))
    }
    fun canKnightMove(): Boolean{
        return ((movedBy == listOf(2, 1)) || movedBy == listOf(1, 2))
    }
    fun canBishopMove(): Boolean{
        return movedBy[0] == movedBy[1]
    }
    fun canQueenMove(): Boolean{
        return (canRookMove() || canBishopMove())
    }
    fun canKingMove(): Boolean{
        return ((movedBy == listOf(0, 1)) || (movedBy == listOf(1, 0)) || (movedBy == listOf(1, 1)))
    }
    return when(piece.toUpperCase()){
        "R" -> canRookMove()
        "K" -> canKingMove()
        "B" -> canBishopMove()
        "Q" -> canQueenMove()
        else -> canKnightMove()
    }
}

fun getFromSq(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): String{
    val piece = move[0]
    val allLocations = pieceView[piece.toString()]
    val possibleLocs = mutableListOf<String>()
    val toSq = move.slice(move.length-2 until move.length)
    if (allLocations != null) {
        for (sq in allLocations){
            if (checkMove(piece.toString(), sq, toSq)){
                possibleLocs.add(sq)
            }
        }
    }
    if (possibleLocs.size == 1){
        return possibleLocs[0]
    }
    val limitPossibleLocs = mutableListOf<String>()
    if (!"nN".contains(piece)) {
        for (sq in possibleLocs) {
            if (notBlocked(sq, toSq, boardView)) {
                limitPossibleLocs.add(sq)
            }
        }
        if (limitPossibleLocs.size == 1) {
            return limitPossibleLocs[0]
        } else {
            return removeAmbiguity(move, limitPossibleLocs)
        }
    }
    else{
        return removeAmbiguity(move, possibleLocs)
    }
}

fun movePiece(move: String, boardView: HashMap<String, String>, pieceView: HashMap<String, MutableList<String>>): Board {
    val fromSq = getFromSq(move, boardView, pieceView)
    val toSq = move.slice(move.length - 2 until move.length)
    if (move.contains('x')) {
        pieceView[boardView[toSq]]!!.remove(toSq)
    }
    pieceView[move[0].toString()]!!.add(toSq)
    pieceView[move[0].toString()]!!.remove(fromSq)
    boardView[fromSq] = SPACE
    boardView[toSq] = move[0].toString()

    return Board(boardView, pieceView)
}